### PR TITLE
Add desktop file generator script

### PR DESCRIPTION
**Pull request by ikstream**
Signed-off-by: Stefan Venz <stefan.venz@protonmail.com>

### CODE REVIEW

**Comment by ikstream** [comment]
To generate the desktop files clone this repo and switch to branch.

Create files for all packages

cd scripts
bash gen-desktop-files -a
Create desktop file for a single package

cd scripts
bash gen-desktop-files -p ../packages/<package>/PKGBUILD 
Create desktop file with icon name different from package name

cd scripts
bash gen-desktop-files -p ../packages/<package>/PKGBUILD  -i <name of icon without png ending>
-a and -p will check if a .png exist in the package directory and add it, if available

**Comment by ikstream** [comment]
Added `-n, --no-version-bump` flag to enable use of the script during new package creation

**Review by Edu4rdSHL — CHANGES_REQUESTED** [review]
Add the suggestions that are suggested in https://www.shellcheck.net/

**Comment by ikstream** [comment]
`-i` option copies the provided file to package dir and rename it to `$packagename.png`

**Review by Edu4rdSHL — APPROVED** [review]
Hey @ikstream, awesome! LGTM, I will do some testing and then will merge into master if no issues are found.

**Comment by ikstream** [comment]
Any updates?

**Comment by Edu4rdSHL** [comment]
I have detected the following issue:

When I run the script directly in the PKGBUILD directory, the file is named in an unexpected way. For example:

```
$ cd ~/BlackArch/blackarch/packages/amass
$ bash gen-desktop-files -p PKGBUILD
```

The `.desktop` file is called `ba-..desktop`, maybe we can take the name from the PKGBUILD `pkgname` variable instead of relying on the directory PATH? Also, the `Name` and `Icon` got the `.` value. Here's the resulting file:

```
[Desktop Entry]
Name="."
Icon="."
Comment="In-depth subdomain enumeration written in Go."
TryExec=/usr/bin/Amass
Exec=sh -c '/usr/bin/"Amass";$SHELL'
StartupNotify=true
Terminal=true
Type=Application
Categories=" X-Blackarch-Scanner; X-Blackarch-Recon;"
```

**Comment by ikstream** [comment]
Thanks for the feedback, will look into it and make it work from the PKGBUILD dir

**Comment by ikstream** [comment]
The script uses the `pkgname` of a package now for the `Name` entry in the desktop file.

In addition I added the `-P, --path` flag to provide a path to the packages directory containing the sub-directories with PKGBUILD files. This does only work with -a | --all and -l | --letter flags. 

Example:
To generate desktop files for all `air*` packages from `blackarch` git root run 
```
bash gen-desktop-files -l air -P packages
```

File: scripts/gen-desktop-files [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+#   pkg: PKGBUILD file
+#   name: desktop/icon file name
+#
+add_source()
+{
+  pkg="$1"
+
+  # if no name is provided, add ba-$pkgname to PKGBUILD file
+  if [ -z "$2" ]; then
+    name='ba-$pkgname.desktop'
```
Comment (Edu4rdSHL) [inline scripts/gen-desktop-files:244]: Expressions don't expand in single quotes, use double quotes for that.
```suggestion
    name="ba-$pkgname.desktop"
```

File: scripts/gen-desktop-files [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+
+# Generate ba-*.desktop files for all packages
+#
+# Arguments:
+#   path - path to PKGBUILD files
+generate_all()
+{
+  path="$1"
+
+  for pkg in $(awk '/groups/ {print FILENAME}' "${path}"/*/PKGBUILD)
```
Comment (Edu4rdSHL) [inline scripts/gen-desktop-files:389]:
```suggestion
  awk '/groups/ {print FILENAME}' "${path}"/*/PKGBUILD |  while IFS= read -r pkg
```

**Review by Edu4rdSHL — CHANGES_REQUESTED** [review]

File: scripts/gen-desktop-files (hunk 1 above)
Comment (ikstream) [reply scripts/gen-desktop-files:244]: It's the intention, that it is not expanded here, but used in PKGBUILD file

Comment (Edu4rdSHL) [reply scripts/gen-desktop-files:244]: Ok, please review the other suggestions as well.

File: scripts/gen-desktop-files [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+#
+# Arguments:
+#   pkg : PKGBUILD file
+#
+write_desktop_file()
+{
+  pkg="$1"
+  name="$(get_name "$pkg")"
+  comment="$(get_desc "$pkg")"
+  ex_name="$(get_prog_name "$pkg")"
```
Comment (Edu4rdSHL) [inline scripts/gen-desktop-files:353]:
```suggestion
  pkg="$1"
  name="$(get_prog_name "$pkg")"
  comment="$(get_desc "$pkg")"
  ex_name="$(get_name "$pkg")"
```

File: scripts/gen-desktop-files [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+      icon="utilities-terminal"
+    else
+      icon="$name"
+    fi
+  else
+    cp "$ICON" "$(dirname "$pkg")/${name}.png"
+    icon="$name"
+  fi
+
+  cat > "$(dirname "$pkg")/ba-${name}.desktop" << EOF
```
Comment (Edu4rdSHL) [inline scripts/gen-desktop-files:366]:
```suggestion
  cat > "$(dirname "$pkg")/ba-${ex_name}.desktop" << EOF
```

**Review by Edu4rdSHL — CHANGES_REQUESTED** [review]
Currently, the package binary path is wrong if the PKGBUILD contains a `$_pkgname` as it takes the value from there, the usage of the functions were the other way around. Also, to preserve lowercase `.desktop` filenames we need to point the name to `$ex_name` instead of `$name`

File: scripts/gen-desktop-files [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+  if [ -z "$ICON" ]; then
+    if [ ! -f "$(dirname "$pkg")/${name}.png" ]; then
+      icon="utilities-terminal"
+    else
+      icon="$name"
+    fi
+  else
+    cp "$ICON" "$(dirname "$pkg")/${name}.png"
+    icon="$name"
+  fi
```
Comment (Edu4rdSHL) [inline scripts/gen-desktop-files:364]:
```suggestion
  if [ -z "$ICON" ]; then
    if [ ! -f "$(dirname "$pkg")/${ex_name}.png" ]; then
      icon="utilities-terminal"
    else
      icon="$ex_name"
    fi
  else
    cp "$ICON" "$(dirname "$pkg")/${ex_name}.png"
    icon="$ex_name"
  fi
```

**Review by Edu4rdSHL — CHANGES_REQUESTED** [review]

**Comment by Edu4rdSHL** [comment]
I forgot to mention, we also need to check if the `ba-$pkgname.desktop` path has been added already to the PKGBUILD, otherwise, we end with something like:

```
source=('git+https://github.com/OWASP/Amass.git'
        "ba-$pkgname.desktop"
        "amass.png"
        "ba-$pkgname.desktop"
        "amass.png"
        "ba-$pkgname.desktop"
        "amass.png"
        "ba-$pkgname.desktop"
        "amass.png")
```

**Comment by ikstream** [comment]
The script will now check if there already is a ba-*.desktop file                   
present. If the -f flag is not provided, the script will terminate.                 
                                                                                    
If you want to force an overwrite, because e.g. you want to provide                 
an icon for the application use                                                     
                                                                                    
```                                                                                 
-f, --force                                                                         
```                                                                                 
                                                                                    
this will update existing shasums in the PKGBUILD file                              
                                                                                    
No duplicate entries should be made in the install section or at source

**Comment by ikstream** [comment]
@noraj do you have any comments? Otherwise I would merge the script.

**Comment by noraj** [comment]
> @noraj do you have any comments? Otherwise I would merge the script.

Nop, the idea looks nice. But I don't use a BA menu.

**Pull request merged**